From: Mel Gorman <mgorman@suse.de>
Date: Fri, 2 Dec 2011 09:50:13 +0000
Subject: [PATCH] mm: Avoid putting a bad page back on the LRU fix
References: 415829
Patch-mainline: no

With commit [57fc4a5e: mm: when migrate_pages returns 0, all pages must
have been released] we can infinite loop on migrating a page with a
memory error. Fix it.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/migrate.c |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/mm/migrate.c b/mm/migrate.c
index c9a54a9..84aba97 100644
--- a/mm/migrate.c
+++ b/mm/migrate.c
@@ -797,6 +797,8 @@ move_newpage:
  		 * restored.
  		 */
  		list_del(&page->lru);
+		dec_zone_page_state(page, NR_ISOLATED_ANON +
+				page_is_file_cache(page));
 		if (PageMemError(page)) {
 			if (rc == 0)
 				/*
@@ -804,7 +806,7 @@ move_newpage:
 				 * been migrated will not be moved to
 				 * the LRU.
 				 */
-				goto move_newpage;
+				goto drop_ref;
 			else
 				/*
 				 * The page failed to migrate and will not
@@ -816,11 +818,10 @@ move_newpage:
 				ClearPageMemError(page);
 		}
 
-		dec_zone_page_state(page, NR_ISOLATED_ANON +
-				page_is_file_cache(page));
 		putback_lru_page(page);
 	}
 
+drop_ref:
 	/*
 	 * Move the new page to the LRU. If migration was not successful
 	 * then this will free the page.
